### PR TITLE
Mat 2094 charging waveforms

### PR DESCRIPTION
**Pull request by patrickherring-TRI**
This PR adds ability to generate waveform charging protocols
- Method to generate stepwise and smooth charging waveforms
- Converter to create the mwf file from the current vs. time waveform
- Generate the procedure and insert the waveform into the correct step
rapid_charge_soc
rapid_charge_matching

Note: Updating the method to include clipping for currents that are too high or too low for the smooth profile. Also, changing the parameterization scheme to three charging rates and two soc points.

### CODE REVIEW

File: beep/protocol/maccor.py
Comment (patrickherring-TRI) [inline beep/protocol/maccor.py:1402]: @chirranjeevigopal-TRI initial soc

File: beep/protocol/maccor.py
Comment (patrickherring-TRI) [inline beep/protocol/maccor.py:719]: @patrickherring-TRI Fix the args

File: beep/protocol/maccor.py
Comment (patrickherring-TRI) [inline beep/protocol/maccor.py:1370]: @patrickherring-TRI Update comments (equally distributed windows over the charging range)

File: beep/utils/waveform.py
Comment (patrickherring-TRI) [inline beep/utils/waveform.py:233]: @pasinger Should this be `mesh_points[-1]`? If so, then some of the smooth curves change substantially when we do the time matching to 80%

Comment (pasinger) [reply beep/utils/waveform.py:248]: I don't think it should be `mesh_points[-1]`. This value is a part of the routine we had decided on for how to set the C rates in the first SOC window. There was some ambiguity originally because we have the C rates "clamping" at the mid-points of the SOC windows, and it would be unclear what behavior should be set for the first C rate before it's mid-point. Obtaining the C rate value from `mesh_points[1]` is part of the routine to either a) set the slope to constant between the first two SOC points or b) set the C rate before the first SOC mid-point to be CC at its specified value. These are calculated in `charging_c_rate_start` below. Case (a) occurs if the slope will be negative and Case (b) occurs if the slope would have been positive.
It may be worth revisiting this behavior since we have added constraints into the new parameterization.

**Review by pasinger — CHANGES_REQUESTED** [review]
Overall looks really good to me. Thanks for implementing all of this!
There are a few pretty minor changes that should be made before approval, and at least one aspect we should review / give further thought to as well.

File: beep/utils/waveform.py
Comment (pasinger) [inline beep/utils/waveform.py:160]: Change line to
charging_c_rates = [charging_c_rates[0]] + list(np.add(charging_c_rates[1:-1],offset)) + [charging_c_rates[-1]]
This change makes it so that only the non-edge c_rate values are offset.

**Review by pasinger — CHANGES_REQUESTED** [review]
Make additional modifications to algorithm around the offsetting to ensure desirable behavior.

File: beep/utils/waveform.py
Comment (patrickherring-TRI) [reply beep/utils/waveform.py:280]: @pasinger I am having some trouble with this change. A bunch of tests are failing since it can't seem to get to the same charge time...

File: beep/protocol/maccor.py
Comment (patrickherring-TRI) [reply beep/protocol/maccor.py]: Done

File: beep/protocol/maccor.py
Comment (patrickherring-TRI) [reply beep/protocol/maccor.py]: Changed this name in the spreadsheet and in the code.

File: beep/protocol/maccor.py
Comment (patrickherring-TRI) [reply beep/protocol/maccor.py]: Changed this back to "charge_current_param_4" to match the other values

File: beep/utils/waveform.py
Comment (patrickherring-TRI) [reply beep/utils/waveform.py]: good catch!

File: beep/utils/waveform.py
Comment (patrickherring-TRI) [reply beep/utils/waveform.py]: Good catch, thanks!

File: beep/utils/waveform.py
Comment (patrickherring-TRI) [reply beep/utils/waveform.py]: changed this one too

File: beep/utils/waveform.py
Comment (patrickherring-TRI) [reply beep/utils/waveform.py:249]: Yeah, I think we can leave the behavior as is for now. I can't really think of a good alternative off-hand.

File: beep/utils/waveform.py
Comment (chirranjeevigopal-TRI) [inline beep/utils/waveform.py:90]: Can we make `soc_points` a global/config variable (if it is not going to be passed as an argument to the constructor)?

File: beep/utils/waveform.py
Comment (chirranjeevigopal-TRI) [inline beep/utils/waveform.py:108]: Would it help to assert that `mesh_points` (soc_list) is monotonically increasing?

File: beep/protocol/maccor.py
Comment (chirranjeevigopal-TRI) [inline beep/protocol/maccor.py:1372]: Minor comment: what's the reason for `v1` designation for this method?

File: beep/utils/waveform.py
Comment (patrickherring-TRI) [reply beep/utils/waveform.py:90]: I don't think that making it a global variable is a good idea here since we might need to alter the interpolation. Is there a specific reason you were thinking it should be a global variable?

File: beep/protocol/maccor.py
Comment (patrickherring-TRI) [reply beep/protocol/maccor.py:1372]: This is a specific way of parameterizing the fast waveform. Since there might be others down the line, I am hedging a bit to allow for new functions that allow for different parameterizations. Usually for these functions, we want legacy functions to continue to work in the same way so that we can reproduce older results.

File: beep/utils/waveform.py
Comment (patrickherring-TRI) [reply beep/utils/waveform.py:108]: Good idea, adding that in

File: beep/utils/waveform.py
Comment (chirranjeevigopal-TRI) [reply beep/utils/waveform.py:90]: No strong reason -- I wasn't sure if having a hardcoded variable buried in was better than having it defined as a config/global variable when we needed to change. But, since its a class variable, I suppose its ok!

**Review by chirranjeevigopal-TRI — APPROVED** [review]
Thanks for making these significant additions. LGTM!

**Review by pasinger — APPROVED** [review]
looks good to me!

**Review by BruisVanVlijmen2020 — APPROVED** [review]
Trusting your wisdom

**Review by pasinger — APPROVED** [review]
looks good!

**Pull request merged**
